cpp/unreal_plugins/SpEngine/Source/SpEngine/WorkQueue.h

```diff
@@ -0,0 +1,125 @@
+//
+// Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+
+#pragma once
+
+#include <concepts>    // std::same_as
+#include <future>
+#include <map>
+#include <mutex>
+#include <string>
+#include <utility>     // std::forward, std::move
+#include <vector>
+#include <type_traits> // std::invoke_result_t
+
+#include <SpCore/Assert.h>
+#include <SpCore/Boost.h>
+
+template <typename TFunc>
+concept CFuncIsCallableWithNoArgs = requires(TFunc func) {
+    { func() };
+};
+
+template <typename TFunc, typename... TArgs>
+concept CFuncIsCallableWithArgs = requires(TFunc func, TArgs... args) {
+    { func(args...) };
+};
+
+template <typename TFunc, typename TReturn, typename... TArgs>
+concept CFuncReturnsAndIsCallableWithArgs = requires(TFunc func, TArgs... args) {
+    { func(args...) } -> std::same_as<TReturn>;
+};
+
+class WorkQueue {
+
+public:
+    WorkQueue() : io_context_(), executor_work_guard_(io_context_.get_executor()) {}
+
+    void run()
```

Comment:
```
// called from the game thread
void run()

// called from a worker thread
void reset()
```

Reply:
Leave these comments in the header, even as we move the implementations into `WorkQueue.cpp`